tracing/src/main/java/com/palantir/tracing/DetachedSpan.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import com.google.errorprone.annotations.MustBeClosed;
+import com.palantir.tracing.api.SpanType;
+
+/**
+ * Span operation which is not bound to thread state, and can measure operations which
+ * themselves aren't bound to individual threads.
+ */
+public interface DetachedSpan {
+
+    /**
+     * Equivalent to {@link Tracer#startSpan(String, SpanType)}, but using this {@link DetachedSpan}
+     * as the parent instead of thread state.
+     */
+    @MustBeClosed
+    SpanToken attach(String operationName, SpanType type);
```

Comment:
What's the difference between calling `detach(operation)` and then `complete()` vs `attach(operation)` and then `close()`? is it just that calling attach also overrides my current thread local state?

Reply:
attach applies a _new_ span to the current thread, from `attach()` to `close()` This can be done multiple times to create sibling spans with the `DetachedSpan` as a parent.

`DetachedSpan.complete`, is used to complete the detached span, which is not associated with a thread. There is danger that complete is never called, and we fail to associate the child spans (from `attach`) with the original parent of this detached span, but I don't think there are good options to avoid that.

This model means that you cannot ever detach or attach an existing span that is already bound to a thread, each time that we attach or detach we mark a new operation. I've found that otherwise, it can become difficult to trace where our tracing spans have actually come from.

Reply:
Oh I see, so you'd start your `DetachedSpan`, then you could `attach()` it once your async task starts running, and then go back to using the normal `Tracer.startSpan()` methods throughout.

Think the thing that seems a bit funky is that you need to close the `SpanToken` rather than just using `Tracer.completeSpan()`.

Reply:
That is correct, the reason for the different type is so we can reset the original tracing data on the attached thread. We could ignore that, and apply state directly, but it would be confusing to api consumers whether or not the attach creates a new tracing span, and how many they are responsible for completing.

Reply:
I think this may have been more obvious to me if the `attach()` method was on `Tracers` rather than here. In my head, at least at the moment, `Tracers` deals with all the threading business, and so I could imagine wanting to call `Tracers.attach(detachedSpan)` which would then give me a `SpanToken` to close and reset the thread's trace. And then on `DetachedSpan` you just would just have `newSpan()` which gives you a new `DetachedSpan` parented by the current one. That way `DetachedSpan` has nothing to do with threading.

Reply:
👍 That sounds good to me.

Reply:
As a point of reference, that also seems to be what these guys are doing as well: https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/Tracer.java#L140

Reply:
`Tracers.attach(detachedSpan)` would also need an operation name param. It would also require us to pass some additional data between DetachedSpan and Tracer, which isn't currently possible. Right now, the DetachedSpan interface doesn't leak whether or not it is sampled, but that _is_ required in order to attach to a thread.